src/boot/efi/util.c

```diff
@@ -769,6 +770,126 @@ EFI_STATUS make_file_device_path(EFI_HANDLE device, const char16_t *file, EFI_DE
+
+/* detect direct kernel boot, i.e. 'qemu -kernel systemd-bootx64.efi' */
+BOOLEAN is_qemu_direct_kernel(EFI_HANDLE device) {
+        EFI_STATUS err;
+        VENDOR_DEVICE_PATH *dp;
+
+        err = BS->HandleProtocol(device, &DevicePathProtocol, (void **) &dp);
+        if (err != EFI_SUCCESS)
+                return FALSE;
+
+        if (dp->Header.Type != MEDIA_DEVICE_PATH ||
+            dp->Header.SubType != MEDIA_VENDOR_DP)
+                return FALSE;
+
+        if (memcmp(&dp->Guid, &QEMU_KERNEL_LOADER_FS_MEDIA_GUID, sizeof(EFI_GUID)) != 0)
+                return FALSE;
+
+        Print(L"loaded via qemu direct kernel boot\n");
+        return TRUE;
+}
+
+/* detect loading from firmware volume (sd-boot added to ovmf) */
+BOOLEAN is_firmware_volume(EFI_HANDLE device) {
+        EFI_STATUS err;
+        VENDOR_DEVICE_PATH *dp;
+
+        err = BS->HandleProtocol(device, &DevicePathProtocol, (void **) &dp);
+        if (err != EFI_SUCCESS)
+                return FALSE;
+
+        if (dp->Header.Type != MEDIA_DEVICE_PATH ||
+            dp->Header.SubType != MEDIA_PIWG_FW_VOL_DP)
+                return FALSE;
+
+        Print(L"loaded from firmware volume\n");
+        return TRUE;
+}
+
+/* try find ESP when not loaded from ESP */
+EFI_STATUS qemu_open(EFI_HANDLE *qemu_dev, EFI_FILE **qemu_dir) {
+        _cleanup_free_ EFI_HANDLE *handles = NULL;
+        _cleanup_free_ CHAR16 *dp_str = NULL;
+        _cleanup_free_ CHAR16 *fs_str = NULL;
+        _cleanup_(file_closep) EFI_FILE *root_dir = NULL;
+        _cleanup_(file_closep) EFI_FILE *efi_dir = NULL;
+        EFI_DEVICE_PATH *dp = NULL;
+        EFI_DEVICE_PATH *fs = NULL;
+        CHAR16 order_str[20];
+        UINTN n_handles = 0;
+        UINTN order = 0;
+        EFI_STATUS err;
+        uint32_t item, dp_len;
+
+        // ovmf doesn't connect devices for direct kernel boot
+        reconnect();
+
+        // find all file system handles
+        err = BS->LocateHandleBuffer(ByProtocol, &FileSystemProtocol, NULL, &n_handles, &handles);
+        if (err != EFI_SUCCESS)
+                return err;
+
+        do {
+                free(dp_str); dp_str = NULL;
+
+                SPrint(order_str, sizeof(order_str), L"QemuBootOrder%04x", order);
+                err = efivar_get_raw(&QEMU_BOOT_ORDER_GUID, order_str, (char**)(&dp), NULL);
+                if (err == EFI_SUCCESS) {
+                        // use device path to filter file systems
+                        dp_str = DevicePathToStr(dp);
+                        dp_len = StrLen(dp_str);
+                } else {
+                        // check all file systems
+                        dp_str = NULL;
+                        dp_len = 0;
+                }
+
+                for (item = 0; item < n_handles; item++) {
+                        file_closep(&efi_dir); efi_dir = NULL;
+                        file_closep(&root_dir); root_dir = NULL;
```

Comment:
Hm, `file_closep()` should clear the pointer.

Here, please declare `efi_dir` and `root_dir` in the loop with cleanup attribute.